platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/ReservedSignedState.java

```diff
@@ -0,0 +1,182 @@
+     * reservation has, as long as that ID is unique.
+     */
+    private static final AtomicLong nextReservationId = new AtomicLong(0);
+
+    /**
+     * Create a wrapper around null (for scenarios where we are storing a null signed state).
+     */
+    public ReservedSignedState() {
+        this.signedState = null;
+        this.reason = "";
+    }
+
+    /**
+     * Create a new reserved signed state.
+     *
+     * @param signedState the signed state to reserve
+     * @param reason      a short description of why this SignedState is being reserved. Each location where a
+     *                    SignedState is reserved should attempt to use a unique reason, as this makes debugging
+     *                    reservation bugs easier.
+     */
+    ReservedSignedState(@NonNull final SignedState signedState, @NonNull final String reason) {
+        this.signedState = Objects.requireNonNull(signedState);
+        this.reason = Objects.requireNonNull(reason);
+
+        // It is safe to "leak this" here.
+        // All fields are final, this class is final, and everything has been instantiated.
+        signedState.incrementReservationCount(this);
+    }
+
+    /**
+     * Check if the signed state is null.
+     *
+     * @return true if the signed state is null, false otherwise
+     */
+    public boolean isNull() {
+        throwIfClosed();
+        return signedState == null;
+    }
+
+    /**
+     * Check if the signed state is not null. If this method returns true then it is not strictly required to call
+     * {@link #close} on this object.
+     *
+     * @return true if the signed state is not null, false otherwise
+     */
+    public boolean isNotNull() {
+        throwIfClosed();
+        return signedState != null;
+    }
+
+    /**
+     * Get another reservation on the signed state.
+     *
+     * @param reason a short description of why this SignedState is being reserved. Each location where a SignedState is
+     *               reserved should attempt to use a unique reason, as this makes debugging reservation bugs easier.
+     * @return a new wrapper around the state that holds a new reservation
+     */
+    public @NonNull ReservedSignedState getAndReserve(@NonNull final String reason) {
+        throwIfClosed();
+        if (signedState == null) {
+            return new ReservedSignedState();
+        }
+        return new ReservedSignedState(signedState, reason);
+    }
+
+    /**
+     * Get the signed state. Does not take any reservations on the signed state, this is the responsibility of the
+     * caller. Do not keep a reference to this signed state outside the scope of this wrapper object without properly
+     * reserving it.
+     *
+     * @return the signed state
+     * @throws NullPointerException if the signed state is null
+     */
+    public @NonNull SignedState get() {
+        throwIfClosed();
+        return Objects.requireNonNull(signedState);
```

Comment:
I would like to see a more concrete IllegalStateException here

Reply:
done